nanofi/include/sitetosite/CSiteToSite.h

```diff
@@ -0,0 +1,423 @@
+
+  UT_hash_handle hh;
+} CTransaction;
+
+static void InitTransaction(CTransaction * transaction, TransferDirection direction, cstream * stream) {
+  transaction->_stream = stream;
+  transaction->_state = TRANSACTION_STARTED;
+  transaction->_direction = direction;
+  transaction->_dataAvailable = 0;
+  transaction->current_transfers_ = 0;
+  transaction->total_transfers_ = 0;
+  transaction->_bytes = 0;
+  transaction->_crc = 0;
+
+  CIDGenerator gen;
+  gen.implementation_ = CUUID_DEFAULT_IMPL;
+  generate_uuid(&gen, transaction->_uuid_str);
+  transaction->_uuid_str[36]='\0';
+}
+
+static TransferDirection getDirection(const CTransaction * transaction) {
+  return transaction->_direction;
+}
+
+static TransactionState getState(const CTransaction * transaction) {
+  return transaction->_state;
+}
+
+static int isDataAvailable(const CTransaction * transaction) {
+  return transaction->_dataAvailable;
+}
+
+static void setDataAvailable(CTransaction * transaction, int available) {
+  transaction->_dataAvailable = available;
+}
+
+static uint64_t getCRC(const CTransaction * transaction) {
+  return transaction->_crc;
+}
+
+static const char * getUUIDStr(const CTransaction * transation) {
+  return transation->_uuid_str;
+}
+
+static void updateCRC(CTransaction * transaction, const uint8_t *buffer, uint32_t length) {
+  transaction->_crc = crc32(transaction->_crc, buffer, length);
+}
+
+static int writeData(CTransaction * transaction, const uint8_t *value, int size) {
+  int ret = write_buffer(value, size, transaction->_stream);
+  transaction->_crc = crc32(transaction->_crc, value, size);
+  return ret;
+}
+
+static int readData(CTransaction * transaction, uint8_t *buf, int buflen) {
+  //int ret = transaction->_stream->read(buf, buflen);
+  int ret = read_buffer(buf, buflen, transaction->_stream);
+  transaction->_crc = crc32(transaction->_crc, buf, ret);
+  return ret;
+}
+
+static int is_little_endian() {
+  static unsigned int x = 1;
+  static char *c = (char*) &x;
+  return  (*c == 1) ? 1 : 0;
+}
+
+static int write_uint64t(CTransaction * transaction, uint64_t base_value) {
+  const uint64_t value = is_little_endian() == 1 ? htonll_r(base_value) : base_value;
+  const uint8_t * buf = (uint8_t*)(&value);
+
+  return writeData(transaction, buf, sizeof(uint64_t));
+}
+
+static int write_uint32t(CTransaction * transaction, uint32_t base_value) {
+  const uint32_t value = is_little_endian() == 1 ? htonl(base_value) : base_value;
+  const uint8_t * buf = (uint8_t*)(&value);
+
+  return writeData(transaction, buf, sizeof(uint32_t));
+}
+
+static int write_uint16t(CTransaction * transaction, uint16_t base_value) {
+  const uint16_t value = is_little_endian() == 1 ? htons(base_value) : base_value;
+  const uint8_t *buf = (uint8_t *) (&value);
+
+  return writeData(transaction, buf, sizeof(uint16_t));
+}
+
+static int write_UTF_len(CTransaction * transaction, const char * str, size_t len, enum Widen widen) {
+  if (len > 65535) {
```

Comment:
This reflects some of the changes in MINIFICPP-676 - payload doesn't get copied during transfer.